Support proposed kid:N markers in cache manager reports

kinkie opened the pull request:
Add support for SMP cache manager report format currently proposed at
https://github.com/squid-cache/squid/pull/1784

kinkie committed:
Support YAML kids reporting

Squid PR 1784 modifies SMP reports to use YAML.
Update overlord to also accept that reporting style

rousskov commented on overlord.pl:
```diff
@@ -391,8 +391,8 @@ sub squidHasAllKids
-    # how many kids completed their by kidN {...} by kidN reports
-    my $kidsRegistered = () = $mgrPage =~ /^[}] by kid\d+/mg;
+    # how many kids completed their per-kid reports
+    my $kidsRegistered = () = $mgrPage =~ /^([}] by kid\d+|\s+kid: \d+)/mg;
```
Thank you for keeping squid-overlord in sync with anticipated Squid changes!

The proposed squid-overlord change violates the "completed" part of the existing regex definition. Counting _completed_ kid reports is required for detecting bugs that crash a reporting kid in the middle of its non-aggregated report generation.

How about counting completed reports using document end marker(s) instead? Squid PR will need to be adjusted as well, of course. There is already a [discussion](https://github.com/squid-cache/squid/pull/1784#discussion_r1569451431) about that change.

kinkie replied:
As of Squid commit 92e101e829, end-of-kid is identified by
... # kid: <number>

It is valid yaml (I checkeD) and fulfills the objective here

kinkie committed:
count kids at end instead of beginning

kinkie committed:
only match ... for end-of-report

rousskov committed:
fixup: Improve regex character quoting consistency

rousskov reviewed (approved):
@kinkie, I pushed a polishing change without fully testing it. Please check 5c6b9d60f64e6da4f35b68de816603de0c88682d.

After your verification, we could merge and deploy this PR, but I hesitate doing so until there are signs that Amos is OK with the proposed `...` document termination approach.

kinkie commented:
> @kinkie, I pushed a polishing change without fully testing it. Please check [5c6b9d6](https://github.com/measurement-factory/squid-overlord/commit/5c6b9d60f64e6da4f35b68de816603de0c88682d).
> 
> After your verification, we could merge and deploy this PR, but I hesitate doing so until there are signs that Amos is OK with the proposed `...` document termination approach.

No rush, we can wait for a few days

rousskov commented on overlord.pl:
```diff
@@ -391,8 +391,8 @@ sub squidHasAllKids
-    # how many kids completed their by kidN {...} by kidN reports
-    my $kidsRegistered = () = $mgrPage =~ /^[}] by kid\d+/mg;
+    # how many kids completed their per-kid reports
+    my $kidsRegistered = () = $mgrPage =~ /^([}] by kid\d+|[.]{3})/mg;
```
As of now, this PR is not sufficient to accommodate current https://github.com/squid-cache/squid/pull/1784 code (as of that PR commit [95f24ece](https://github.com/squid-cache/squid/pull/1784/commits/95f24ece51217ea65d4c80cce53694c53f71a9aa)) because the latter PR also affects kidsRegistered vs. kidsExpected math by adding `by kid` framing to non-SMP Squids, increasing kidsRegistered values for those Squids. The changes in non-SMP kidsRegistered values trigger a die() assertion a few lines below.

I believe I know what code changes are needed, but they will need to be done in both Daft and Squid Overlord repositories. I am not yet sure whether we can easily and safely support both current and post-1784 Squids with the same Daft/Overlord code. I hope to avoid this, but it may be necessary to require other Squid PRs to incorporate PR 1784 changes after that PR merges into Squid master/v7.

rousskov replied:
> The changes in non-SMP kidsRegistered values trigger a die() assertion a few lines below.

Fixed in branch commit 8a835fbc.

rousskov committed:
Support POP v9 with its worker-count and disker-count options

Drop support for the kids-expected option.

Anticipated Squid changes[^1] introduce "by kid" framing to non-SMP
cache manager reports. This framing changes the number of expected
kid-specific sections in those reports. We could accommodate those
changes by adjusting squidHasAllKids() math, but that would break
existing/older Squids. To support both Squid variations, Overlord now
receives raw worker/disker counts and computes the number of sections.

Anticipated Squid changes[^1] also replace "by kid" framing with YAML
document markers (in YAML-compliant reports). Adjusted $kidSections
regex now accommodates both grammar variations.

[^1]: https://github.com/squid-cache/squid/pull/1784 (as of 4dbf863).
